File: Backend/mapping/src/test/java/com/mhp/coding/challenges/mapping/ApplicationTests.java
package com.mhp.coding.challenges.mapping;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;

import static org.hamcrest.Matchers.containsString;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@RunWith(SpringRunner.class)
@SpringBootTest
@AutoConfigureMockMvc
public class ApplicationTests {

    @Autowired
    private MockMvc mockMvc;

    // acceptance criterion 1 - returns json data
    @Test
    public void shouldReturnAllArticles() throws Exception {
        this.mockMvc.perform(get("/article"))
                .andDo(print())
                .andExpect(status().isOk())
                .andExpect(content().string(containsString("Article Description 1001")))
                .andExpect(content().string(containsString("Article Description 2002")))
                .andExpect(content().string(containsString("Article Description 3003")))
                .andExpect(content().string(containsString("Article Description 4004")))
                .andExpect(content().string(containsString("Article Description 5005")));
    }

    // acceptance criterion 1 - returns json data
    @Test
    public void shouldReturnSingleArticle() throws Exception {
        this.mockMvc.perform(get("/article/1001"))
                .andDo(print())
                .andExpect(status().isOk())
                .andExpect(content().string(containsString("Article Description 1001")));
    }

    // acceptance criterion 3 - returns 404 for unknown id
    @Test
    public void shouldReturnNotFoundForUnknownId() throws Exception {
        this.mockMvc.perform(get("/article/9999"))
                .andDo(print())
                .andExpect(status().isNotFound());
    }

}

File: Backend/mapping-kotlin/src/test/kotlin/com/mhp/coding/challenges/mapping/ApplicationTests.kt
package com.mhp.coding.challenges.mapping

import com.mhp.coding.challenges.mapping.models.dto.ArticleDto
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.boot.test.web.client.getForEntity
import org.springframework.http.HttpStatus.OK

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class ApplicationTests {

    @Autowired
    lateinit var testRestTemplate: TestRestTemplate

    @Test
    fun contextLoads() {
    }

    @Test
    fun `controller should return article list`() {
        val result = testRestTemplate.getForEntity<List<ArticleDto>>("/article")

        assertThat(result.statusCode).isEqualTo(OK)

        assertThat(result.body).hasSize(5)
    }

    @Test
    fun `controller should return article by id`() {
        val result = testRestTemplate.getForEntity<String>("/article/1")

        assertThat(result.statusCode).isEqualTo(OK)

        assertThat(result.body).contains("Article Description 1")
    }
}
